ql/src/java/org/apache/hadoop/hive/ql/schq/ScheduledQueryExecutionService.java

```diff
@@ -0,0 +1,159 @@
+
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.QueryState;
+import org.apache.hadoop.hive.metastore.api.ScheduledQueryPollResponse;
+import org.apache.hadoop.hive.metastore.api.ScheduledQueryProgressInfo;
+import org.apache.hadoop.hive.ql.DriverFactory;
+import org.apache.hadoop.hive.ql.IDriver;
+import org.apache.hadoop.hive.ql.processors.CommandProcessorResponse;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+public class ScheduledQueryExecutionService implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScheduledQueryExecutionService.class);
+
+  private ScheduledQueryExecutionContext context;
+  private ScheduledQueryExecutor worker;
+
+  public static ScheduledQueryExecutionService startScheduledQueryExecutorService(HiveConf conf) {
+    MetastoreBasedScheduledQueryService qService = new MetastoreBasedScheduledQueryService(conf);
+    ExecutorService executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Scheduled Query Thread %d").build());
+    ScheduledQueryExecutionContext ctx = new ScheduledQueryExecutionContext(executor, conf, qService);
+    return new ScheduledQueryExecutionService(ctx);
+  }
+
+  public ScheduledQueryExecutionService(ScheduledQueryExecutionContext ctx) {
+    context = ctx;
+    ctx.executor.submit(worker = new ScheduledQueryExecutor());
+    ctx.executor.submit(new ProgressReporter());
+  }
+
+  class ScheduledQueryExecutor implements Runnable {
+
+    private ScheduledQueryProgressInfo info;
+
+    @Override
+    public void run() {
+      while (true) {
+        ScheduledQueryPollResponse q = context.schedulerService.scheduledQueryPoll();
+        if (q.isSetExecutionId()) {
+          try{
+            processQuery(q);
+          } catch (Throwable t) {
+            LOG.error("Unexpected exception during scheduled query processing", t);
+          }
+        } else {
+          try {
+            Thread.sleep(context.getIdleSleepTime());
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.info("interrupted");
+            break;
+          }
+        }
+      }
+    }
+
+    QueryState lastState;
+    public synchronized void reportQueryProgress() {
+      if (info != null) {
+        if(lastState == QueryState.FINISHED || lastState == QueryState.ERRORED) {
+          return;
+        }
+        LOG.info("Reporting query progress of {} as {} err:{}", info.getScheduledExecutionId(), info.getState(),
+            info.getErrorMessage());
+        context.schedulerService.scheduledQueryProgress(info);
+        lastState=info.getState();
+      }
+    }
+
+    private void processQuery(ScheduledQueryPollResponse q) {
```

Comment:
What happens in the following two scenarios with current implementation?
1) A scheduled query is modified while it is being executed.
2) HS2 becomes irresponsive for a certain time, the grab on the lock expires, another HS2 grabs the query execution, the first HS2 comes back to life.

Reply:
We need to fix 2. One of the common use cases that we were discussing for scheduled queries was incrementally refreshing contents in a table in a transactional way, e.g., pulling data from Kafka. Currently, if query can be assumed dead and then come back to life, we may basically end up inserting data more than once in target object. You could just grab an exclusive execution lock on the scheduled query identifier to prevent this from happening, we already have mechanisms to do that (fwiw we have similar logic for incremental materialized view rebuild).

Reply:
I don't think that this issue should be in the scope of scheduled queries - I understand your point; but I don't think it should be fixed here...I mean by scheduled queries...

* If I launch a query in a hs2 
* before it finishes I pause the instance
* and then run some similar query in another instance
* ...when I unpause the first one ; how it supposed to work correctly and not write its results?

I believe the above is a separate problem